text/0000-command-ergonomics.md

```diff
@@ -0,0 +1,542 @@
+}
+struct SubprocessError { ... }
+impl From<SubprocessError> for io::Error { ... }
+```
+
+The `.output()` function and `std::process::Output`
+will be deprecated.
+
+No significant changes are made to the `Command` construction APIs,
+but it may become necessary to call `.stderr()` explicitly
+to use the new methods (see Unresolved Questions).
+
+## Use cases
+
+We aim to serve well each of the following people:
+
+ * Alice writes a CLI utility to orchestrate processes and wants top-notch error reporting from subprocesses.
+
+ * Bob migrates ad-hoc automation scripts from bash to Rust.
+
+ * Carol writes a generic application and just wants all errors to be maximally useful by default.
+
+ * Dionysus wants to run `diff`, which exits `0` for "no difference",
+   `1` for "difference found" and
+   another value for failure.
+
+(Partly cribbed from one of
+@matklad's [comments](https://github.com/rust-lang/rust/pull/89004#issuecomment-923803209) in #89004)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## New methods on `Command`:
+
+ * `fn run(&mut self) -> Result<(), SubprocessError>`:
+
+   Runs the command.
+   Equivalent to `.spawn()` followed by `.status()`,
+   but with better error handling.
+
+ * `fn get_output_bytes(&mut self) -> Result<Vec<u8>, SubprocessError>`:
+
+   Runs the command and collects its stdout.
+   After the child indicates EOF on its stdout,
+   we will wait for it to finish and check the exit status.
+
+ * `fn get_output(&mut self) -> Result<Vec<String>, SubprocessError>`:
+
+   Runs the command and collects its stdout.
+   Decodes the stdout as UTF-8, and fails if that's not possible.
+   Does not trim any trailing line ending.
+
+ * `fn get_output_line(&mut self) -> Result<Vec<String>, SubprocessError>`:
+
+   Runs the command and collects its stdout.
+   Decodes the stdout as UTF-8, and fails if that's not possible.
+   Fails unless the output is a single line (with or without line ending).
+   Trims the line ending (if any).
+
+ * `fn get_output_read(&mut self) -> std::process::ChildOutputStream`
+   (where `struct ChildOutputStream` implements `io::Read`
+   and is `Send + Sync + 'static`).
+
+   Starts the command, allowing the caller to
+   read the stdout in a streaming way.
+   Neither EOF nor an error will be reported by `ChildOutputStream`
+   until the child has exited, *and* the stdout pipe reports EOF.
+   (This includes errors due to nonempty stderr,
+   if stderr was set to `piped`.)
+
+Most callers should use these methods,
+rather than `.spawn()` or `.status()`.
+These new methods should be be
+recommended by docs for other, lower-level functions
+which people were previously required to use.
+
+## Deprecations
+
+ * Deprecate `std::process::Command::output()`.
+   This API cannot be fixed;
+   see [#73126](https://github.com/rust-lang/rust/issues/73126).
+
+ * Apply `#[must_use]` to `Command` and `Child`.
```

Comment:
:tada: :tada: :tada:

Reply:
Would it be possible to make this change without an RFC? Seems somewhat unfortunate to tie this together with the API additions.